### DISCUSSION TITLE
How to get global value inside DebugHandler (step/break functions)

### DISCUSSION THREAD

**Discussion by lkathke** [sebastienros/jint#940] [Q&A]
As the title is showing,
I would like to know, how I can get global/local values while I am inside the step/break function of the DebugHandler.

Release: 3.0.0-beta-2033

js-code:
```
var o = { 'Hallo': 'Welt', 'Test': 0};
o = { 'Hallo': 'Welt2', 'Test': 1};
o = { 'Hallo': 'Welt3', 'Test': 2};
```

c# code:
```
var engine = new Engine(c => { 
                c.AllowClr(typeof(WebClient).Assembly);
                c.DebugMode();
                c.DebuggerStatementHandling(Jint.Runtime.Debugger.DebuggerStatementHandling.Script);               
            });

var handler = engine.DebugHandler;
handler.Step += Handler_Step;
handler.BreakPoints.Add(new Jint.Runtime.Debugger.BreakPoint(1, 0));

....

 private static Jint.Runtime.Debugger.StepMode Handler_Step(object sender, Jint.Runtime.Debugger.DebugInformation e)
 {
      var f = e.CurrentCallFrame.ScopeChain.Global.GetBindingValue("o");
 }
```
e.CurrentCallFrame.ScopeChain.Global.GetBindingValue("o") doesn't return the object's value (it returns "undefined"). Do I have to bind them first somewhere?

**Comment by lahma** [comment]
Converted to discussion and pinging @Jither for possible ideas and insights.

**Comment by Jither** [comment]
Testing on 2033, your code works fine for me. Note, however, that your breakpoint doesn't actually call the event handler on `Step`, which is only triggered when stepping - to stop at the breakpoint, you'd need to add an event handler on the `Break` event. And at that break, the variable is, indeed, `undefined` (`1, 0` = first line, first character - lines are 1-based in Esprima, and Jint - so this is before any code runs)

Slightly modified code:

```
void Main()
{
	var engine = new Engine(c =>
	{
		c.AllowClr(typeof(WebClient).Assembly);
		c.DebugMode();
		c.DebuggerStatementHandling(Jint.Runtime.Debugger.DebuggerStatementHandling.Script);
	});
	
	var handler = engine.DebugHandler;
	handler.Break += Handler_Break;
	handler.Step += Handler_Step;
	handler.BreakPoints.Add(new Jint.Runtime.Debugger.BreakPoint(1, 0));
	engine.Execute(@"var o = { 'Hallo': 'Welt', 'Test': 0};
o = { 'Hallo': 'Welt2', 'Test': 1};
o = { 'Hallo': 'Welt3', 'Test': 2};");
}

private static StepMode Handler_Step(object sender, DebugInformation e)
{
	var f = e.CurrentCallFrame.ScopeChain.Global.GetBindingValue("o");
	Console.WriteLine($"Step o: {f}");
	var prop = f.Get("Hallo");
	Console.WriteLine($"Step: {prop}");
	return StepMode.Into;
}

private static StepMode Handler_Break(object sender, DebugInformation e)
{
	var f = e.CurrentCallFrame.ScopeChain.Global.GetBindingValue("o");
	Console.WriteLine($"Break o: {f}");
	var prop = f.Get("Hallo");
	Console.WriteLine($"Break: {prop}");
	return StepMode.Into;
}
```

... outputs:

```
Break o: undefined
Break: undefined
Step o: [object Object]
Step: Welt
Step o: [object Object]
Step: Welt2
```

**Reply by lkathke** [reply]
This works for me, thank you for the quick help!! 👌 @Jither @lahma

**Reply by Jither** [reply]
By the way, for local values, it's a bit more involved, since the `Local` property *only* refers to the local "`var`" scope inside functions. The debugger follows pretty much the same conventions as Chromium's debugger, so depending on what you need, you may need to iterate through multiple scopes to find the value you're looking for (`ScopeChain` is an `IReadOnlyList` with the `Local` and `Global` properties just being (not always *that*) convenient shortcuts). The types of scopes are probably most usefully documented in https://github.com/sebastienros/jint/blob/main/Jint/Runtime/Debugger/DebugScopeType.cs
